sstables/sstable_directory.cc

```diff
@@ -558,65 +558,64 @@ bool sstable_directory::compare_sstable_storage_prefix(const sstring& prefix_a,
-                // do is to prepend/drop the trailing slash, it should be enough to
-                // compare prefixes of both ... prefixes
-                assert(compare_sstable_storage_prefix(first->_storage->prefix(), sst->_storage->prefix()));
-            }
+            auto toc = sst->component_basename(component_type::TOC);
+            co_await out->write(toc);
+            co_await out->write("\n");
         }
 
-        sstring pending_delete_dir = first->_storage->prefix() + "/" + sstables::pending_delete_dir;
-        sstring pending_delete_log = format("{}/sstables-{}-{}.log", pending_delete_dir, gen_tracker.min(), gen_tracker.max());
-        sstring tmp_pending_delete_log = pending_delete_log + ".tmp";
-        sstlog.trace("Writing {}", tmp_pending_delete_log);
-        try {
-            touch_directory(pending_delete_dir).get();
-            auto oflags = open_flags::wo | open_flags::create | open_flags::exclusive;
-            // Create temporary pending_delete log file.
-            auto f = open_file_dma(tmp_pending_delete_log, oflags).get();
-            // Write all toc names into the log file.
-            auto out = make_file_output_stream(std::move(f), 4096).get();
-            auto close_out = deferred_close(out);
-
-            for (const auto& sst : ssts) {
-                auto toc = sst->component_basename(component_type::TOC);
-                out.write(toc).get();
-                out.write("\n").get();
-            }
+        co_await out->flush();
+        co_await out->close();
+        out.reset();
 
-            out.flush().get();
-            close_out.close_now();
+        dir_f = co_await open_directory(pending_delete_dir);
+        // Once flushed and closed, the temporary log file can be renamed.
+        co_await rename_file(tmp_pending_delete_log, pending_delete_log);
+        ret_pending_delete_log = &pending_delete_log;
 
-            auto dir_f = open_directory(pending_delete_dir).get();
-            auto close_dir = deferred_close(dir_f);
-            // Once flushed and closed, the temporary log file can be renamed.
-            rename_file(tmp_pending_delete_log, pending_delete_log).get();
+        // Guarantee that the changes above reached the disk.
+        co_await dir_f->flush();
+        co_await dir_f->close();
+        dir_f.reset();
+        sstlog.debug("{} written successfully.", pending_delete_log);
+    } catch (...) {
+        ex = std::current_exception();
+    }
 
-            // Guarantee that the changes above reached the disk.
-            dir_f.flush().get();
-            close_dir.close_now();
-            sstlog.debug("{} written successfully.", pending_delete_log);
-        } catch (...) {
-            sstlog.warn("Error while writing {}: {}. Ignoring.", pending_delete_log, std::current_exception());
+    if (ex) {
+        sstlog.warn("Error while writing {}: {}. Ignoring.", pending_delete_log, std::current_exception());
```

Comment:
We are out of try-catch block. Is usage of `std::current_exception()` valid here? Shouldn't we use `ex` variable?

From stackoverflow [[1]](https://stackoverflow.com/questions/68841864/is-it-allowed-to-call-stdcurrent-exception-in-lambda-which-is-called-in-a-ca):

> An exception handler is active after its catch-clause argument is initialized and until the catch clause argument has just been destroyed. If there is not currently handled by an active handler an exception, std:current_exception will return an empty std::exception_ptr.


Toy-example which indicates, that usage of `std::current_exception` results in empty message:
https://godbolt.org/z/cvKsc4bnc

Reply:
good catch!  should `s/std::current_exception()/ex/`

Reply:
fixed in v2